In [1]:
Nom_Cognom_1 = "David Morillo Massagué"
NIU_1        = "1666540"
Nom_Cognom_2 = "Adrià Muro Gómez"
NIU_2        = "1665191"

# Pràctica 1: Codificació de Huffman.


La Cristina, professora d'Informació i Seguretat, ha perdut la vostra nota de pràctiques del primer laboratori. Us demana que li passeu la vostra nota de pràctiques pel fòrum de l'assignatura amb el missatge $mNota=$_"Som NIU1 , NIU2, la nota es X"_. On $X$, com encara no sabeu la vostra nota suposem que és la suma dels últims dígits de cada NIU mòdul 10. (Per exemple: si el NIU del Pau és 1567897 i el de la Maria és  163899, la nota serà  $(7+9)\mod 10 = 6$. En aquest cas el missatge seria $mNota=$_"Som 1567897 , 163899, la nota es 6"_).     

Per culpa del ciberatac de la UAB, el campus virtual encara no està operatiu. De moment us esteu comunicant per un fòrum molt limitat. Per assegurar-se que tots els missatges hi caben al fòrum, us demana que li envieu el missatge en binari i comprimit segons la codificació decidida pel vostre professor de pràctiques. El vostre professor de pràctiques és un gran seguidor de David Huffman i té implementada una funció secreta que només compartirà amb vosaltres si resoleu els següents dos reptes:

**1.** Implementeu la funció $generarProbabilitats(missatge)$ que retorni un diccionari on les claus són els diferents símbols del missatge i el valor la probabilitat amb què apareix el símbol dins el missatge. Doneu el resultat d'aplicar la funció a la cadena $mNota$ anterior. _[1.5 pts]_
    Quin símbol té la probabilitat més gran?_[0.25 pts]_ A què es deu aquest fet? _[0.25 pts]_.

In [2]:
### Implementeu aquí el vostre codi
mNota = "Som 1665191 , 1666540, la nota es 1"
total = len(mNota)
def generarProbabilitats(missatge):
    dic = {}
    for simbol in missatge:
      if simbol not in dic.keys():
        dic[simbol] = 1

      else:
        dic[simbol] += 1

    for simbol in dic.keys():
      dic[simbol] = dic[simbol]/total

    return dic

dictProbabilitats = generarProbabilitats(mNota)

**Resposta:** Els caràcters més probables son l'espai " " i "1" (freqüència: 1/5). Ens basem en aquesta afirmació degut a que observem que és el caràcter que és vegades apareix al missatge i, estadísticament podem dir que és el que té més probabilitats d'aparèixer.

**2.** Implementeu la funció $entropia(dictProbabilitats)$ on dictProbabilitats és un diccionari que té com a claus els símbols i valors les seves probabilitats. Quina és l'entropia del missatge $mNota$? _[1 pts]_

In [3]:
from math import log2

### Implementeu aquí el vostre codi
def entropia(dictProbabilitats):
    entropia = 0
    for p in dictProbabilitats.values():
      entropia += p*log2(p)

    return -entropia

#entropia(dictProbabilitats)

**Resposta:** El valor de l'entropia del missatge és de: 3.4920553338362095

Enhorabona!!! Heu superat els dos reptes i ja pertanyeu al selecte club de seguidors/es de Huffman i podem compartir amb vosaltres la funció secreta; s'anomena $codiHuffman()$. Com seguidors/es has d'experimentar la importància de la codificació de Huffman.
Seguiu els següents passos per entendre-la.

**3.** Utilitzeu la funció $convertiraBinari(missatge)$ per convertir el missatge $mNota$ a bits. Quina és la longitud del missatge en bits? _[0.5 pts]_

In [4]:
#Retorna en un array on cada posició es un símbol del missatge codificat en bits.
def convertiraBinari(a):
  l,m=[],[]
  for i in a:
    l.append(ord(i))
  for i in l:
    m.append(int(bin(i)[2:]))
  return m

### Implementeu aquí el vostre codi
def longBits(missatge):
  binari = convertiraBinari(missatge)
  #print(len(binari*8))

  long = 0
  for i in binari:
    long += len(str(i))

  return long

longBits(mNota)

221

**Resposta:** 221 bits amb els caràcters en binari donats per la funció convertiraBinari() i sense separar caràcters.

**4.** Seguidament, utilitzeu la llibreria Huffman de SageMath per generar el codi de Huffman binari.
    Completeu la funció $longitudMissatgeCodificat(missatge)$ i doneu el diccionari de codificació _[1.5 pts]_.

In [5]:
from sage.coding.source_coding.huffman import Huffman

#Multiplica els valors de dos diccionaris. Fixa't que dict1 te strings com a valors i dict2 te numeros.
def multiplicarValorsDiccionari(dict1, dict2):
    dict3 = dict(dict1)
    for i, j in dict1.items():
        for x, y in dict2.items():
            if i == x:
                dict3[i]=(len(j)*y)
    return dict3

#Completeu la implementació
def longitudMissatgeCodificat(missatge):
    codificacio = Huffman(missatge).encoding_table()
    frequencies = generarProbabilitats(missatge)
    n = len(missatge)
    frequencies = {x: y*n for x, y in frequencies.items()} # transformem a valors absoluts per cada caràcter
    bitsSymbolsMessage = multiplicarValorsDiccionari(codificacio, frequencies)
    return sum(bitsSymbolsMessage.values())

longitudMissatgeCodificat(mNota)

124

**5.** Quina longitud en bits obtens del missatge $mNota$? _[0.5 pts]_ Surt a compte enviar el missatge codificat pel fòrum? _[0.25 pts]_

**Resposta:** Longitud: 124 bits. Veiem que surt a compte enviar el missatge codificat fent servir Huffman

diccionari de Huffman: {'S': '10100', 'e': '10101', 'l': '10110', 'm': '10111', 'n': '11000', 's': '11001', 't': '11010', ',': '11011', '5': '0110', '9': '0111', 'a': '1000', 'o': '1001', '6': '010', ' ': '111', '1': '00'}


Atès que ja heu arribat fins aquí, el professor de pràctiques que està acabant d'escriure la seva tesi doctoral, necessita un cop de mà de seguidors/es de Huffman per entregar la tesi a temps. A més si l'ajudeu us promet que ho reflectirà en la vostra nota de pràctiques del primer laboratori.  

**6.** Implementeu la funció $longitudCodiHuffman(missatge)$ que retorni la longitud mitjana del codi de Huffman binari generat pel missatge. Quina és la longitud del codi generat pel missatge $mNota$? _[1.5 pts]_

In [6]:
### Implementeu aquí el vostre codi
def longitudCodiHuffman(missatge):
    len_bits = longitudMissatgeCodificat(missatge)
    len_simbols = len(missatge)
    return len_bits / len_simbols

longitudCodiHuffman(mNota)

124/35

**Resposta:** 124/35 bits per caràcter (aprox. 3,54)

**7.** Implementeu la funció $redundanciaCodiHuffman(missatge)$ que retorna la redundància del codi Huffman generat per un missatge. Quina és la redundància del codi de Huffman generat pel missatge $mNota$? _[1.5 pts]_

In [7]:
### Implementeu aquí el vostre codi
def redundanciaCodiHuffman(missatge):
    
    diccionari = generarProbabilitats(missatge)
    ent = entropia(diccionari)
    l_mitjana = longitudCodiHuffman(missatge)
    redundancia = 1 - (ent/(l_mitjana*log2(len(diccionari.items()))))
    
    return redundancia

redundanciaCodiHuffman(mNota)

0.7477122136924566

**Resposta:** Redundància: 0.7477122136924566 

**8.** Un codi qualsevol amb la redundància anterior podem considerar que és òptim? _[0.5 pts]_ Podem trobar un altre codi binari associat al missatge $mNota$ amb longitud mitjana menor que l'obtinguda a l'apartat 4? _[0.75 pts]_  (Justifica degudament les respostes)

**Resposta:** 

A) És complicat decidir si amb la redundància anterior (0.7477122136924566) es podria considerar el codi com a òptim. Aquest valor indica que aproximadament el 74% de les dades en la seqüència original encara estan presents en la seqüència comprimida.  En altres paraules, encara hi ha una quantitat significativa de dades innecessàries o repetitives en la representació comprimida. Encara que considerem que s'ha comprimit bastant, creiem que té bastant marge de millora, en termes de compressió, i no el considerem del tot òptim.

B) Seria possible trobar altres codis binaris pel missatge amb longitud menor a la que ens proporciona l'algoritme de Huffman, ja que aquest ens proporciona uns codis basat en les freqüències calculades amb el missatge introduit, que és relativament curt. Si sapiguéssim en detall les probabilitats de cada caràcter sense basar-nos en les simples freqüències del missatge, en tot cas es podria millorar, teòricament.

La Cristina i el/la professor/a de pràctiques agraïm la vostra ajuda i el vostre esforç per restablir la nota que us mereixeu. Esperem que el campus virtual es recuperi de l'atac informàtic i puguem penjar la vostra nota de pràctiques del primer laboratori al més aviat possible. Esperem que no se'ns torni a perdre :)